### ISSUE TITLE
Unexpected problem when computing LOO for a Poisson model

### ISSUE THREAD

**Issue by EmmanuelCharpentier** [paul-buerkner/brms#689]
TL;DR : I can't compute LOO for *one* **specific** (simple) model where other quite analogous models deriving from the same data are fine.

Devil's resting place (i. e. details : 

R setup:

``` r
library(survival)
library(ggplot2)
library(rstan)
options(mc.cores = parallel::detectCores())
rstan_options(auto_write = TRUE)
library(bayesplot)
library(brms)
library(digest)
```

Exemple data stolen from the [Web page](https://data.princeton.edu/wws509/notes/c7s5) of an (interesting) Princeton course: infant mortality in Colombia in three different period, tabulated by age:

``` r
Colombie <-
structure(list(Âge = structure(c(1L, 2L, 3L, 4L, 5L, 6L, 7L, 
1L, 2L, 3L, 4L, 5L, 6L, 7L, 1L, 2L, 3L, 4L, 5L, 6L, 7L), .Label = c("0–1", 
"1–3", "3–6", "6–12", "12–24", "24–60", "60–120"), class = "factor"), 
    Décès = c(168L, 48L, 63L, 89L, 102L, 81L, 40L, 197L, 48L, 
    62L, 81L, 97L, 103L, 39L, 195L, 55L, 58L, 85L, 87L, 70L, 
    10L), Exposition = c(278.4, 538.8, 794.4, 1550.8, 3006, 8743.5, 
    14270, 403.2, 786, 1165.3, 2294.8, 4500.5, 13201.5, 19525, 
    495.3, 956.7, 1381.4, 2604.5, 4618.5, 9814.5, 5802.5), Cohorte = structure(c(1L, 
    1L, 1L, 1L, 1L, 1L, 1L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 3L, 3L, 
    3L, 3L, 3L, 3L, 3L), .Label = c("1941–59", "1960–67", 
    "1968–76"), class = "factor"), tEnd = c(1, 3, 6, 12, 24, 
    60, 120, 1, 3, 6, 12, 24, 60, 120, 1, 3, 6, 12, 24, 60, 120
    )), row.names = c(NA, -21L), class = "data.frame")
digest(Colombie)
```

The point is to model infant mortality in Columbia according to age and to study cohort ; the choosen mode is a piecewise-constant, proportional hazard mode. Using the usual "Poisson trick", this leads to the straightforward saturated model:

``` r
C1 <- local({
  T <- brm(Décès ~ 0+Âge*Cohorte+offset(log(Exposition)),
     data=Colombie, family=poisson(link=log),
     prior=prior(student_t(3, 0, 10), class="b"),
     save_all_pars=TRUE, sample_prior="yes",
     silent=TRUE, open_progress=FALSE, seed=1789,
     control=list(adapt_delta=0.99, max_treedepth=15))
  add_criterion(T, crit=c("loo", "marglik"), reloo=TRUE, silent=TRUE,
      open_progress=FALSE, overwrite=TRUE,
      model_name="C1")
})
digest(C1)
```

Variation of this model can be built by:

  - omitting the Age\*Cohort interaction (i. e. modeling the effects of age and cohort as independent):

``` r
CAC <- local({ 
  T <- brm(Décès ~ 0+Âge+Cohorte+offset(log(Exposition)),
     data=Colombie, family=poisson(link=log),
     prior=prior(student_t(3, 0, 10), class="b"),
     save_all_pars=TRUE, sample_prior="yes",
     silent=TRUE, open_progress=FALSE, seed=1789,
     control=list(adapt_delta=0.99, max_treedepth=15))
  add_criterion(T, crit=c("loo", "marglik"), reloo=TRUE, silent=TRUE,
      open_progress=FALSE, overwrite=TRUE,
      model_name="CAC")
})
digest(CAC)
```

  - omitting the cohort (i. e. modeling the gross effect of age):

``` r
CA <- local({
  T <- brm(Décès ~ 0+Âge+offset(log(Exposition)),
     data=Colombie, family=poisson(link=log),
     prior=prior(student_t(3, 0, 10), class="b"),
     save_all_pars=TRUE, sample_prior="yes",
     silent=TRUE, open_progress=FALSE, seed=1789,
     control=list(adapt_delta=0.99, max_treedepth=15))
  add_criterion(T, crit=c("loo", "marglik"), reloo=TRUE, silent=TRUE,
      open_progress=FALSE, overwrite=TRUE,
      model_name="CA")
})
digest(CA)
```

  - omitting the age (i. e. modeling the gross effect of cohort):

``` r
CC <- local({
  T <- brm(Décès ~ 0+Cohorte+offset(log(Exposition)),
     data=Colombie, family=poisson(link=log),
     prior=prior(student_t(3, 0, 10), class="b"),
     save_all_pars=TRUE, sample_prior="yes",
     silent=TRUE, open_progress=FALSE, seed=1789,
     control=list(adapt_delta=0.99, max_treedepth=15))
  add_criterion(T, crit=c("loo", "marglik"), reloo=TRUE, silent=TRUE,
      open_progress=FALSE, overwrite=TRUE,
      model_name="CC")
})
digest(CC)
```

All of these variations work fine: I get sensible results (quite close to those given by `glm`, as expected with such weakly informative priors, sensible model comparisons by `loo` or `bayes_factor`), **except for the last model.**

Further research shows that the problem arises when trying to compute `LOO` for the last model. Fitting the model and computing its marginal likelihood wokrs fine:

``` r
foo <- brm(Décès ~ 0+Cohorte+offset(log(Exposition)),
     data=Colombie, family=poisson(link=log),
     prior=prior(student_t(3, 0, 10), class="b"),
     save_all_pars=TRUE, sample_prior="yes",
     silent=TRUE, open_progress=FALSE, seed=1789,
     control=list(adapt_delta=0.99, max_treedepth=15))
bar  <- add_criterion(foo, crit=c("marglik"), reloo=TRUE, silent=TRUE,
            open_progress=FALSE, overwrite=TRUE,
            model_name="CC")
summary(bar$marglik)
```

| Logml<sub>Estimate</sub> | Relative<sub>MeanSquaredError</sub> | Coefficient<sub>ofVariation</sub> | Percentage<sub>Error</sub> | Method | Repetitions |
| ------------------------ | ----------------------------------- | --------------------------------- | -------------------------- | ------ | ----------- |
| \-2176.55193837868       | 1.4286053474329e-06                 | 0.00119524279852794               | 0.120%                     | normal | 1           |

But requesting `loo` fails:

``` r
gee  <- add_criterion(bar, crit=c("loo"), reloo=TRUE, silent=TRUE,
          open_progress=FALSE, overwrite=TRUE,
          model_name="CC")
```

``` example

Erreur : Can't mix NA and not NA values in 'r_eff'.
```

I am unable to understand this error message. May it be e bug somewhere in `brms` ?

**Comment by paul-buerkner** [comment]
I think I have fixed this already. Can you try it again with the latest dev version of brms?

**Comment by EmmanuelCharpentier** [comment]
And, by the way : 
```r
> vignette(package="brms")
no vignettes found
```

Ouch...

**Comment by paul-buerkner** [comment]
Very likely there are some extreme outliers driving this difference. You
can likely see this in the loo output of the model under question.

EmmanuelCharpentier <notifications@github.com> schrieb am Di., 18. Juni
2019, 17:46:

**Comment by paul-buerkner** [comment]
When installing the dev version by default no vignettes are built.

Paul Buerkner <paul.buerkner@gmail.com> schrieb am Di., 18. Juni 2019,
17:53:

**Comment by EmmanuelCharpentier** [comment]
Dear Paul;

1. The 0-1 month age has indeed extreme mortality (a fact well known to any pediatrician/neonatalogist).

2. Is there a workaround to force vignette installation when installing from Github ? If any, the doc doesn't mention it...

**Comment by jgabry** [comment]
If you’re using devtools::install_github then you can just use its `build_vignettes` argument. Installation will be much slower though.

**Comment by EmmanuelCharpentier** [comment]
Dear Jonah,

> If you’re using devtools::install_github then you can just use its build_vignettes argument.

Thanks for the tip. However, one should note that this argument is currently *not* documented by `devtools`. BTW, it seems outdated. According to this [devtools issue](https://github.com/r-lib/devtools/issues/1896), one should use `devtools::install_github("paul-buerkner/brms", build_opts="")` in order to build these docs.

> Installation will be much slower though.

Not really a problem. More problematic is the build dependency of vignettes on `R.rsp`, which itself has a bucketful of dependencies...

Note/question to Paul Buerkner : shouldn't this be mentioned on the main page ?

**Comment by paul-buerkner** [comment]
I am not sure. Building all of brms' vignettes may take a few hours and you can easily access vignettes outside of R for instance on CRAN (https://cran.r-project.org/web/packages/brms/index.html) or the brms website (https://paul-buerkner.github.io/brms/articles/index.html). 

I am closing this issue since the original problem has been resolved.

**Comment by EmmanuelCharpentier** [comment]
> Building all of brms' vignettes may take a few hours

Huh ?

```r
system.time(devtools::install_github("paul-buerkner/brms", build_opts="", force=TRUE))
Downloading GitHub repo paul-buerkner/brms@master
   checking for file ‘/tmp/RtmpfVUWqP/remotes4b31ab0157/paul-buerkner-brms-2af54✔  checking for file ‘/tmp/RtmpfVUWqP/remotes4b31ab0157/paul-buerkner-brms-2af5404/DESCRIPTION’
─  preparing ‘brms’:
✔  checking DESCRIPTION meta-information ...
─  installing the package to build vignettes
✔  creating vignettes (28.8s)
─  checking for LF line-endings in source and make files and shell scripts
─  checking for empty or unneeded directories
─  looking to see if a ‘data/datalist’ file should be added
─  building ‘brms_2.9.2.tar.gz’
   
Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)
* installing *source* package ‘brms’ ...
** using staged installation
** R
Registered S3 method overwritten by 'xts':
  method     from
  as.zoo.xts zoo 
** data
*** moving datasets to lazyload DB
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
*** copying figures
** building package indices
** installing vignettes
** testing if installed package can be loaded from temporary location
** testing if installed package can be loaded from final location
** testing if installed package keeps a record of temporary installation path
* DONE (brms)
utilisateur     système      écoulé 
     52.086       2.651      64.368 
```

I am not aware of a location where those vignettes could have been hidden... Weren't them in the source tarball ?

**Comment by paul-buerkner** [comment]
Yes exactly. They need to be evaluated within a certain environmental variable to actually compute results.

```
withr::with_envvar(c("NOT_CRAN" = "true"),
  # installation code here
)
```

**Comment by EmmanuelCharpentier** [comment]
It turns out that the installation time isn *that* bad. on my current laptop : 
```r
> system.time(withr::with_envvar(c("NOT_CRAN"="true"), devtools::install_github("paul-buerkner/brms", build_opts="", build_vignettes=TRUE, force=TRUE)))
Downloading GitHub repo paul-buerkner/brms@master
   checking for file ‘/tmp/RtmpyOKVm4/remotes14084854b68e/paul-buerkner-brms-2af✔  checking for file ‘/tmp/RtmpyOKVm4/remotes14084854b68e/paul-buerkner-brms-2af5404/DESCRIPTION’
─  preparing ‘brms’:
✔  checking DESCRIPTION meta-information ...
─  installing the package to build vignettes
✔  creating vignettes (56m 33.6s)
─  checking for LF line-endings in source and make files and shell scripts
─  checking for empty or unneeded directories
─  looking to see if a ‘data/datalist’ file should be added
─  building ‘brms_2.9.2.tar.gz’
   
* installing *source* package ‘brms’ ...
** using staged installation
** R
Registered S3 method overwritten by 'xts':
  method     from
  as.zoo.xts zoo 
** data
*** moving datasets to lazyload DB
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
*** copying figures
** building package indices
** installing vignettes
** testing if installed package can be loaded from temporary location
** testing if installed package can be loaded from final location
** testing if installed package keeps a record of temporary installation path
* DONE (brms)
utilisateur     système      écoulé 
   4272.229      62.703    3425.665 
```

I still think that this documentation, possibly adorned by a fair warning, could go on the main page. One hour of clock time (possibly dedicated to such frills as sleep or feeding...) isn't much. BTW, the installation of Stan itself is worse...
